src/core/lib/event_engine/posix_engine/posix_endpoint.cc

```diff
@@ -0,0 +1,1135 @@
+    return cmsg;
+  }
+
+  auto tss = reinterpret_cast<scm_timestamping*>(CMSG_DATA(cmsg));
+  auto serr = reinterpret_cast<struct sock_extended_err*>(CMSG_DATA(next_cmsg));
+  if (serr->ee_errno != ENOMSG ||
+      serr->ee_origin != SO_EE_ORIGIN_TIMESTAMPING) {
+    gpr_log(GPR_ERROR, "Unexpected control message");
+    return cmsg;
+  }
+  // The error handling can potentially be done on another thread so we need to
+  // protect the traced buffer list. A lock free list might be better. Using a
+  // simple mutex for now.
+  {
+    absl::MutexLock lock(&traced_buffer_mu_);
+    traced_buffers_.ProcessTimestamp(serr, opt_stats, tss);
+  }
+  return next_cmsg;
+}
+
+void PosixEndpointImpl::HandleError(absl::Status status) {
+  if (!status.ok() ||
+      stop_error_notification_.load(std::memory_order_relaxed)) {
+    // We aren't going to register to hear on error anymore, so it is safe to
+    // unref.
+    Unref();
+    return;
+  }
+  // We are still interested in collecting timestamps, so let's try reading
+  // them.
+  if (!ProcessErrors()) {
+    // This might not a timestamps error. Set the read and write closures to be
+    // ready.
+    handle_->SetReadable();
+    handle_->SetWritable();
+  }
+  handle_->NotifyOnError(on_error_);
+}
+
+bool PosixEndpointImpl::WriteWithTimestamps(struct msghdr* msg,
+                                            size_t sending_length,
+                                            ssize_t* sent_length,
+                                            int* saved_errno,
+                                            int additional_flags) {
+  if (!socket_ts_enabled_) {
+    uint32_t opt = kTimestampingSocketOptions;
+    if (setsockopt(fd_, SOL_SOCKET, SO_TIMESTAMPING, static_cast<void*>(&opt),
+                   sizeof(opt)) != 0) {
+      return false;
+    }
+    bytes_counter_ = -1;
+    socket_ts_enabled_ = true;
+  }
+  // Set control message to indicate that you want timestamps.
+  union {
+    char cmsg_buf[CMSG_SPACE(sizeof(uint32_t))];
+    struct cmsghdr align;
+  } u;
+  cmsghdr* cmsg = reinterpret_cast<cmsghdr*>(u.cmsg_buf);
+  cmsg->cmsg_level = SOL_SOCKET;
+  cmsg->cmsg_type = SO_TIMESTAMPING;
+  cmsg->cmsg_len = CMSG_LEN(sizeof(uint32_t));
+  *reinterpret_cast<int*>(CMSG_DATA(cmsg)) = kTimestampingRecordingOptions;
+  msg->msg_control = u.cmsg_buf;
+  msg->msg_controllen = CMSG_SPACE(sizeof(uint32_t));
+
+  // If there was an error on sendmsg the logic in tcp_flush will handle it.
+  ssize_t length = TcpSend(fd_, msg, saved_errno, additional_flags);
+  *sent_length = length;
+  // Only save timestamps if all the bytes were taken by sendmsg.
+  if (sending_length == static_cast<size_t>(length)) {
+    traced_buffer_mu_.Lock();
+    traced_buffers_.AddNewEntry(static_cast<uint32_t>(bytes_counter_ + length),
+                                fd_, outgoing_buffer_arg_);
+    traced_buffer_mu_.Unlock();
+    outgoing_buffer_arg_ = nullptr;
+  }
+  return true;
+}
+
+#else
```

Comment:
I believe this goes with the `GRPC_LINUX_ERRQUEUE` 230 lines earlier. Please comment here.

Reply:
Add a comment.